src/interfaces.py

```diff
@@ -1,5 +1,105 @@
 from collections.abc import Callable
-from typing import Protocol
+from enum import Enum
+from typing import List, Protocol, Tuple
+
+
+class ISimulationObject(Protocol):
+    """An object synchronized with the simulation"""
+
+    traci_id: str
+
+    def update(self, updates: dict) -> None:
+        """Gets called whenever a simualtion tick has happened.
+
+        :param updates: The updated values for the synchronized properties
+        """
+
+    def add_subscriptions(self) -> int:
+        """Gets called when this object is created to allow
+        specification of simulator-synchronized properties.
+
+        :return: The synchronized properties (see <https://sumo.dlr.de/pydoc/traci.constants.html>)
+        """
+
+
+class ISimulationNode(ISimulationObject, Protocol):
+    """A point somewhere in the simulation where `ISImulationTrack`s meet"""
+
+    position: Tuple[float, float]
+
+
+class ISimulationSignal(ISimulationNode, Protocol):
+    """A signal in the simulation which can either show stop or go
+    (See `ISimulationSignal.State`)
+    """
+
+    class State(Enum):
+        """The possible states of the signal"""
+
+        HALT = 1
+        GO = 2
+
+    state: State
+
+
+class ISimulationSwitch(ISimulationNode, Protocol):
+    """A switch in the simulation which can point either left or right
+    (see `ISimulationSwitch.State`)
+    """
+
+    class State(Enum):
+        """The possible states of the switch"""
+
+        LEFT = 1
+        RIGHT = 2
+
+    state: State
+
+
+class ISimulationTrack(ISimulationObject, Protocol):
+    """A track in the simulation where trains can drive along"""
+
+    blocked: bool
+    speed_limit: float
+
+
+class ISimulationPlatform(ISimulationObject, Protocol):
+    """A platform where trains can arrive, load and unload passengers and depart"""
+
+    track: ISimulationTrack
+    station_name: str
+    platform_number: int
+    blocked: bool
+
+
+class ISimulationTrain(ISimulationObject, Protocol):
+    """A train driving around in the simulation."""
+
+    current_track: ISimulationTrack
+    priority: int
+    position: Tuple[float, float]
+    speed: float
+    max_speed: float
+    route: str
+    timetable: List[ISimulationPlatform]
+    _train_type: str
+
+    def __init__(
```

Comment:
Why does `ISimulationTrain` have a `__init__`function and the other Protocols doesn't? 🤔 Even the `ISimulationObject`...

Reply:
@CR1337 `__init__` should be part of a Protocol, shouldn't it?

Reply:
It is only here because this is the only method that has a implementor-facing creation option (we could implement a class-method `spawn_train` or something instead). In every other object, the `__init__` is part of the private api.

Reply:
On the other hand it is important information how `__init__` works. I would say practicality over dogma and lets have the  `__init__` inside the Protocol when construction is part of the interface?

Reply:
Sounds reasonable 🤔 Then we have to edit some existing Protocols, correct?